### DISCUSSION TITLE
Lua any vs typescript any

### DISCUSSION THREAD

**Discussion by sloonz** [teal-language/tl#691] [Q&A]
With typescript any, you can do anything with an any, and it will just happily return anything : 

```
const x: any;
const y = x.foo; // Will work, y will have any type
const z = x.bar(1, "hello", y) // Will work too, z is still any
```

For dealing with untyped external stuff, I strongly prefer the typescript approach. I would like to achieve the same thing with teal ; here is what I came up with : 

```
global type Any = record
	metamethod __index: function(Any, string): Any
	metamethod __call: function(...: any): Any...
end
```

Which kinda works : 

```
local x: Any
x.foo.bar(1, "hello", x.t).spam.egg() -- tl check OK
```

Weirdly, this doesn't : 

```
local x: Any
x:foo() -- invalid key 'foo' in record 'x' of type Any
```

Is this a bug or am I missing something ?

**Comment by lenscas** [comment]
teal's any is like `unknown` in TS.

I am personally not against having a TS like `any` kind of thing but it should be reserved for interfacing with lua and nowhere else. Even TS has been slowly replacing where `any` is used and replacing them with `unknown`

**Reply by sloonz** [reply]
> but it should be reserved for interfacing with lua and nowhere else

Yeah, that’s my primary use-case. Writing `(((ExternalTable as {string: any}).ExternalFunction) as function(...: any): any...)` gets very tedious very fast.

**Reply by hishamhm** [reply]
The goal is to drive people to write `.d.tl` files with typed definitions for the Lua libraries they're interfacing with, rather than keeping considerable chunks of their code untyped. If you `require("mylib")` which is an untyped Lua library, you can keep that in Lua, but add to Teal's lookup path a file `mylib.d.tl` with only a record definition with the types of the public functions. Your code will still run using the Lua library, but the compiler will typecheck it using the definition file. The community maintains a collection of definition files for popular libraries at https://github.com/teal-language/teal-types.

Having said that, Teal does have a type like the one you expect: it's called `unknown` (yes, the names are flipped when compared to TypeScript...) but currently you can't declare it explicitly from Teal code.

However, if you run Teal in "lax mode", which is the default mode when typechecking files with the `.lua` extension (instead of `.tl`), then any variable that does not have a declared type will automatically be typed `unknown`, and no typechecks happen to them. The intention here was to allow users to take their `.lua` codebase and gradually add types until they're ready to switch to `.tl`.
